hail/src/main/scala/is/hail/cxx/Emit.scala

```diff
@@ -797,6 +797,59 @@ class Emitter(fb: FunctionBuilder, nSpecialArgs: Int, ctx: SparkFunctionContext)
            """.stripMargin,
           resultRegion)
 
+      case ir.MakeNDArray(data, shape, rowMajor) =>
+        val elementPType = data.pType.asInstanceOf[PContainer].elementType
+        val datat = emit(data)
+        val shapet = emit(shape)
+        val rowMajort = emit(rowMajor)
+
+        val elemSize = elementPType.byteSize
+        val elemAlignment = elementPType.alignment
+        val flags = fb.variable("flags", "int", s"${rowMajort.v} ? 1 : 0")
+        val shapeVec = fb.variable("shapeVec", "std::vector<long>",
+          s"load_vector<long, true, 8, 8>(${shapet.v})")
+        val entries = fb.variable("entries", "const char *",
```

Comment:
This can/should be `BaseArrayImpl::elements_address(${ datat.v })`.

Also---there's a method on `PContainer` called `cxxImpl` that will get you the right ArrayImpl to use. While this will give the correct pointer address, `ArrayAddrImpl` is conceptually intended to represent arrays whose values are are byte arrays that are stored inline (currently, structs and tuples), which is not the case here.

There's also the question of what you want to do if the value generated by the emit triplet is an array with missingness (which, currently in the general case, it will be). You're not currently checking for (invalid) missing values in the data array, but the missingness information gets thrown out, which means we're potentially left doing calculation on memory garbage instead of throwing an error.

Reply:
Is the missingness not caught in checking `datat.m` before the variable defines?

Reply:
The missingness of the overall array, yes. The thing that hadsn't been checked is the missingness of each individual data entry, if the array element type isn't required.